### PR TITLE
Downcase authentication keys and humanize error message

### LINKED ISSUE #4095: "Invalid Login or password." is grammatically incorrect
This was changed in PR #4014 and released with 4.1.0.

With an incorrect password the user sees "Invalid Login or password." "Login" shouldn't be capitalized in this case, as it's not a proper noun and isn't at the beginning of the sentence.

This is because of the following i18n string:

``` yaml
invalid: "Invalid %{authentication_keys} or password."
```
#4014 "humanized" the `authentication_keys` variable, thus causing this issue.

### PR DESCRIPTION
**Pull request by grantzau**
To fix #4095

### CODE REVIEW

File: lib/devise/failure_app.rb [hunk 1]
```diff
@@ -103,11 +103,11 @@ def i18n_message(default = nil)
         options[:scope] = "devise.failure"
         options[:default] = [message]
         auth_keys = scope_class.authentication_keys
-        keys = (auth_keys.respond_to?(:keys) ? auth_keys.keys : auth_keys).map { |key| scope_class.human_attribute_name(key) }
+        keys = (auth_keys.respond_to?(:keys) ? auth_keys.keys : auth_keys).map { |key| scope_class.human_attribute_name(key).downcase }
```
Comment (feliperenan) [inline lib/devise/failure_app.rb]: Could we add tests for it? 🤔

Comment (grantzau) [reply lib/devise/failure_app.rb]: Edited the existing tests. They check, that the downsizing are done. There's not really a test to check the added humanize method though, not sure, if that's necessary?

Comment (feliperenan) [reply lib/devise/failure_app.rb]: I would add to ensure that it's the expected behavior.

**Comment by grantzau** [comment]
Great! I will look at it and squash it during the weekend. Was editing
directly on Github from the browser, which wasn't a good idea ..!

fre. 6. apr. 2018 kl. 03.22 skrev Felipe Renan <notifications@github.com>:

File: test/failure_app_test.rb [hunk 2]
```diff
@@ -153,20 +153,30 @@ def call_failure(env_params={})
...
+
+    test 'humanizes the flash message' do
+      call_failure('warden' => OpenStruct.new(message: :invalid))
+      assert_equal @request.flash[:alert], @request.flash[:alert].humanize
+    end
+
     test 'uses custom i18n options' do
       call_failure('warden' => OpenStruct.new(message: :does_not_exist), app: FailureWithI18nOptions)
-      assert_equal 'User Steve does not exist', @request.flash[:alert]
+      assert_equal 'User steve does not exist', @request.flash[:alert]
```
Comment (tegon) [inline test/failure_app_test.rb]: I'm not sure about this one, I think it should have stayed as it was before. This is a custom message defined here: https://github.com/plataformatec/devise/blob/88724e10adaf9ffd1d8dbfbaadda2b9d40de756a/test/support/locale/en.yml#L5 and we're passing the user's name `Steve` in a capitalized form here: https://github.com/grantzau/devise/blob/cca5a7b8e9f699de4299e001d95e3e269470f47f/test/failure_app_test.rb#L27, which gets lost when we call `#humanize`
```ruby
[3] pry(main)> "User Steve does not exist".humanize
=> "User steve does not exist"
```
I'm not sure what's the best solution here. Maybe we should only call `#humanize` when the message is `:invalid` 🤔

**Comment by smoyth** [comment]
Note that you can control i18n of parameters to failure messages by doing something like this:

In your models:
```
module People
  module Users
    # Overriding the i18n_options method in the Devise FailureApp so that we can have control
    # over how the `authentication_keys` parameter to the messages under devise.failure is translated.
    # By default the `human_attribute_name` of the attribute is used, but those are typically capitalized,
    # while the messages call for a lowercase value.
    # This custom failure app must be activated in the config/initializers/devise.rb file.
    class CustomDeviseFailureApp < Devise::FailureApp
      def i18n_options(options)
        options.merge(authentication_keys: I18n.t("devise.authentication_keys.email"))
      end
    end
  end
end
```

and then in `config/initializers/devise.rb`:

```
  config.warden do |manager|
    manager.failure_app = People::Users::CustomDeviseFailureApp
  end
```

and then in your `devise.yml` file:

```
  devise:
    ...
    authentication_keys:
      email: "email"
```

Ideally Devise would support an optional `authentication_keys` key in the YML file out of the box, so you could have control over these params without a custom failure app, but this works right now and is not particularly hacky. There is support for this method in the tests even, see [here](https://github.com/plataformatec/devise/blob/master/test/failure_app_test.rb#L171).

Hope this helps!

**Comment by tegon** [comment]
@grantzau are you still going to work on this?

I'm asking to know if I should assign it to someone else.

**Pull request merged**